codegen/src/definition.rs

```diff
@@ -41,254 +15,211 @@ use syn::{
+                Err(format_err_spanned!(
                     item,
-                    "only methods are supported in the trait definition",
+                    "only methods are supported in trait definitions"
                 ))
             }
-        }
-
-        let mut method_descriptions = vec![];
-        for method in methods {
-            if let Some(default) = method.default {
-                return Err(format_err_spanned!(
-                    default,
-                    "default implementation is not supported in chains extension",
-                ))
-            }
-
-            let config = MethodConfig::new(&method)?;
-            let hash = into_u32(&method.sig.ident);
-            let id = config.id;
-            let input = config.input;
-            let output = config.output;
-            let (impls, types, where_clause) = trait_item.generics.split_for_impl();
-            // TODO: Add check that each `id` is unique
-            method_descriptions.push(quote! {
+        })
+        .try_collect()?;
+
+    let method_descriptions = methods.iter().map(
+        |Method {
+             id,
+             hash,
+             input_tokens,
+             output_tokens,
+             ..
+         }| {
+            quote! {
                 impl #impls ::obce::codegen::MethodDescription<#hash> for dyn #trait_name #types #where_clause {
                     const ID: ::core::primitive::u16 = #id;
-                    type Input = #input;
-                    type Output = #output;
+                    type Input = #input_tokens;
+                    type Output = #output_tokens;
                 }
-            });
-        }
-
-        // Remove all `obce` attributes from trait's methods
-        trait_item.items.iter_mut().for_each(|item| {
-            if let TraitItem::Method(method) = item {
-                method.attrs = method
-                    .attrs
-                    .clone()
-                    .into_iter()
-                    .filter(|attr| !attr.path.is_ident("obce"))
-                    .collect();
             }
-        });
-
-        let id = config.id;
-        let substrate = Self::substrate(trait_item.clone())?;
-        let ink = Self::ink(trait_item.clone(), &config)?;
-        let (impls, types, where_clause) = trait_item.generics.split_for_impl();
-
-        let code = quote! {
-            impl #impls ::obce::codegen::ExtensionDescription for dyn #trait_name #types #where_clause {
-                const ID: ::core::primitive::u16 = #id;
+        },
+    );
+
+    let mut ink_trait_item = trait_item.clone();
+
+    ink_trait_item
+        .items
+        .iter_mut()
+        .zip(methods.iter())
+        .for_each(|(item, method)| {
+            if let TraitItem::Method(method_item) = item {
+                method.fill_with_ink_data(&trait_attrs, method_item);
+            } else {
+                // FIXME: Remove unreachable call by using types?
```

Comment:
what about this?

Reply:
I added a comment that describes why this `unreachable!` call exists. While it can be removed with types, it requires us to create one more method iterator with `filter_map`ed elements, which is meaningless, considering that we filtered methods for `trait_item` already, and `ink_trait_item` is a clone of that.